src/mongo/db/audit/audit.cpp

```diff
@@ -214,6 +227,101 @@ namespace audit {
         SimpleMutex _mutex;
     };    
 
+    // Writes audit events to a json file
+    class JSONAuditLog: public FileAuditLog {
+
+        class Adapter: public AuditLogFormatAdapter {
+            const std::string str;
+
+        public:
+            Adapter(const BSONObj &obj)
+                : str(mongoutils::str::stream() << obj.jsonString() << "\n") {}
+
+            virtual const char *data() const override {
+                return str.c_str();
+            }
+            virtual unsigned size() const override {
+                return str.size();
+            }
+        };
+
+    protected:
+        virtual AuditLogFormatAdapter *getAdapter(const BSONObj &obj) const override {
+            return new Adapter(obj);
+        }
+    public:
+        JSONAuditLog(const std::string &file, const BSONObj &filter)
+            : FileAuditLog(file, filter) {}
+    };
+
+    // Writes audit events to a bson file
+    class BSONAuditLog: public FileAuditLog {
+
+        class Adapter: public AuditLogFormatAdapter {
+            const BSONObj &obj;
+
+        public:
+            Adapter(const BSONObj &aobj)
+                : obj(aobj) {}
+
+            virtual const char *data() const override {
+                return obj.objdata();
+            }
+            virtual unsigned size() const override {
+                return obj.objsize();
+            }
+        };
+
+    protected:
+        virtual AuditLogFormatAdapter *getAdapter(const BSONObj &obj) const override {
+            return new Adapter(obj);
+        }
+
+    public:
+        BSONAuditLog(const std::string &file, const BSONObj &filter)
+            : FileAuditLog(file, filter) {}
+    };
+
+    // Writes audit events to the console
+    class ConsoleAuditLog : public WritableAuditLog {
+    public:
+        ConsoleAuditLog(const BSONObj &filter)
+            : _matcher(filter.getOwned()) {
+        }
+    private:
+        const Matcher _matcher;
```

Comment:
Can you move `_matcher` variable to `WritableAuditLog` and make it private, then make `append` non-virtual in a form of:
```
void append(const BSONObj &obj) {
    if (_matcher.matches(obj)) {
        appendMatched(obj);
    }
}
```
and declare `virtual void appendMatched(const BSONObj &obj) = 0;` so there's less code duplication?